### ISSUE TITLE
[13.0] Performance issue on cancelling stock pickings

### ISSUE THREAD

**Issue by Yenthe666** [odoo/odoo#45838]
Impacted versions: 13.0 (tested) but probably way back too.

Steps to reproduce:
1. Have a database with over 100.000 moves and some big purchase orders (100+ lines)
2. Open a big purchase order (in this example one with 153 PO lines) and open the related receipt
3. Trigger the button `action_cancel` which in return triggers `_action_cancel` from `stock.move` which finally triggers `_do_unreserve` and make sure that you have a profiler bound to it.

Current behavior: The process takes ~502 seconds to process on a similar set as this (140 `stock.move.line` records):
User #2 deleted stock.move.line records with IDs: [23257, 23258, 23259, 23260, 23261, 23262, 23263, 23264, 23265, 23266, 23267, 23268, 23269, 23270, 23271, 23272, 23273, 23274, 23275, 23276, 23277, 23278, 23279, 23280, 23281, 23282, 23283, 23284, 23285, 23286, 23287, 23288, 23289, 23290, 23291, 23292, 23293, 23294, 23295, 23296, 23297, 23298, 23299, 23300, 23301, 23302, 23303, 23304, 23305, 23306, 23307, 23308, 23309, 23310, 23311, 23312, 23313, 23314, 23315, 23316, 23317, 23318, 23319, 23320, 23321, 23322, 23323, 23324, 23325, 23326, 23327, 23328, 23329, 23330, 23331, 23332, 23333, 23334, 23335, 23336, 23337, 23338, 23339, 23340, 23341, 23342, 23343, 23344, 23345, 23346, 23347, 23348, 23349, 23350, 23351, 23352, 23353, 23354, 23355, 23356, 23357, 23358, 23359, 23360, 23361, 23362, 23363, 23364, 23365, 23366, 23367, 23368, 23369, 23370, 23371, 23372, 23373, 23374, 23375, 23376, 23377, 23378, 23379, 23380, 23381, 23382, 23383, 23384, 23385, 23386, 23387, 23388, 23389, 23390, 23391, 23392, 23393, 23394, 23395, 23396, 23397] 

Output from the profiler:
```
calls     queries   ms
stock.move -------------------------- /odoo13/odoo13-server/addons/stock/models/stock_move.py, 605

1         0         0.01          @profile
                                  def _do_unreserve(self):
1         0         0.02              moves_to_unreserve = self.env['stock.move']
143       0         2.12              for move in self:
141       0         3.44                  if move.state == 'cancel':
                                              # We may have cancelled move in an open picking in a "propagate_cancel" scenario.
                                              continue
141       0         3.43                  if move.state == 'done':
                                              if move.scrapped:
                                                  # We may have done move in an open picking in a scrap scenario.
                                                  continue
                                              else:
                                                  raise UserError(_('You cannot unreserve a stock move that has been set to \'Done\'.'))
141       0         19.32                 moves_to_unreserve |= move
1         5613      502139.01             to_unreserve = moves_to_unreserve.mapped('move_line_ids').unlink()
1         0         0.01              return True

```

It looks like the problematic line is taking a lot of time to do the `mapped()` and `unlink()` operations. Adding `with_context(prefetch_fields=False)` seems to save ~7% of the time but is still not an acceptable speed.

Expected behavior:
Way faster times

**Comment by Yenthe666** [comment]
OPW `2201251`

**Comment by sle-odoo** [comment]
Why an issue AND an opw?

**Comment by Yenthe666** [comment]
@sle-odoo ahh yes I had this discussion with an Odoo employee before.
TL;DR info: I log it here so it is public and everyone can access/search on it. I keep track of both the OPW and issue and close this issue once it is fixed (along with links to PR's etc). Don't worry, doesn't cost you any extra time.

**Comment by sle-odoo** [comment]
> Don't worry, doesn't cost you any extra time.

if there's multiple people working on a single ticket (and it is bound to happen if you open opw and github issues) yeah one of them will lose time. anyway i won't go into this flamewar again, let's wait from the support

**Comment by sle-odoo** [comment]
(also, i know it would help the support if you provide them a flamegraph)

**Comment by fmdl** [comment]
it seems you have not the last version, now mapped and unlink are at the end

![image](https://user-images.githubusercontent.com/16716992/74962691-ad2f7580-5410-11ea-90c6-2dff5479ce4d.png)

**Comment by wtaferner** [comment]
@fmdl 
I think Yenthe played around too much with the code, but basically I took a quick look on my side and he checked on his instance (maybe with wrong code, but this would probably give a higher gain than he calculated with the code inside the loop). @Yenthe666 you should better have cherry-picked :wink: 

I was testing only in 11.0 with data from one of our instances and https://github.com/odoo/odoo/pull/45851 (50% fasater) is quite a performance boost for these unlink operations (basically skipping the prefetching of the mapped field which is only needed to unlink)

**Comment by Yenthe666** [comment]
Yeah I forgot to change back one of my changes. I was testing the code lines and where the issues where coming from, that's why :-) I'll try to find some time to provide a pyflame too. Either way, it's slowwww

**Comment by sle-odoo** [comment]
there's other thing at play because unlinking 140 move lines is instant at my end (purchase 140 units of a serial product and cancel the receipt)

**Comment by fmdl** [comment]
some number about mapped perf https://github.com/odoo/odoo/pull/30975

**Comment by wtaferner** [comment]
@fmdl 
I liked your PR a year ago and still like it, even though I forgot about it...yours is on a generic level and probably valid IMO.
So, I guess either yours is finally merged (in 11.0) or my little explicit unlink PR relieve and performance will win on all stable and future versions :-)

**Comment by gdgellatly** [comment]
What I think you want is an EXPLAIN ANALYZE on the SQL. I was just solving this elsewhere as it popped up in pg_stats as a top 5 query DELETE FROM stock_move_line and DELETE FROM stock_move.

Check if you have an index on picking_id in stock_move_line and stock_move.

**Comment by gdgellatly** [comment]
I just checked, these are the indexes I added
On SML
    "manual_stock_move_line_lot_id" btree (lot_id)
    "manual_stock_move_line_picking_id" btree (picking_id)
On stock_move
    "manual_stock_move_origin_returned_move_id" btree (origin_returned_move_id)

**Comment by Yenthe666** [comment]
@gdgellatly this index was actually added two days ago in Odoo itself at https://github.com/odoo/odoo/commit/a36a693733579e1e8bdaa3aa00fcc482c6299acf

**Comment by gdgellatly** [comment]
I dont really keep track, just check worst queries, anything over 1 percent CPU time, optimize, reset, check a week later and repeat. Maybe 11ms to sub millisecond I think.

**Comment by Yenthe666** [comment]
Interesting. So how exactly do you filter the queries that Odoo generates and are slow? And do you manually run `EXPLAIN ANALYZE` on those slow queries afterwards then?

**Comment by gdgellatly** [comment]
```sql
SELECT regexp_replace(substring(query, 1, 100), '\s+', ' ', 'g')  AS short_query,
              round(total_time::numeric, 2) AS total_time,
              calls,
              round(mean_time::numeric, 2) AS mean,
              round((100 * total_time /
              sum(total_time::numeric) OVER ())::numeric, 2) AS percentage_cpu
FROM    pg_stat_statements
ORDER BY total_time DESC
LIMIT 20;
```
Is my goto query. Sometimes I add in min and max time. Sometimes I increase the substring number for really long queries.

Then yeah normally I will start with a \d table and check the existing indexes, BEGIN a transaction, run EXPLAIN ANALYZE or similar, try an index, if it is a massive improvement COMMIT; or else ROLLBACK and start again. 

But really, mostly its pretty good, we are talking only about 15 extra indexes over the course of a year and a lot of the time they will be trigram indexes.  They aren't often the answer (typically shorter length queries on larger tables with few joins, with a long runtime, common user openended ilike searches or popular tree views default filters), sometimes it is reworking python code (lots of IN's, or lots of JOINS usually tells me an onchange or something is doing stupid things), sometimes it is disabling things you don't care about (for me that is usually a translatable field). But you kind of need to really know your data as you end up filling a lot of normalized parameters. and know what queries are coming from automated processes rather than users, know what they were probably trying to do etc.

**Comment by Yenthe666** [comment]
Hmm thanks for the tips @gdgellatly! Looks like there are some very slow ones (not directly related to this specific report). For example opening the warehouse analysis on a database with 84K moves or cancelling a PO with 153 lines:
```
short_query                                                                                                                                                                                                                                                      | total_time | calls |  mean  | percentage_cpu 
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------+-------+--------+----------------
 SELECT "product_product".id FROM "product_template" as "product_product__product_tmpl_id" LEFT JOIN "ir_translation" as "product_product__product_tmpl_id__name" ON ("product_product__product_tmpl_id"."id" = "product_product__product_tmpl_id__name"."res_id" AND "product_product__product_tmpl_id__name"."type" = $1 AND "product_product__product_tmpl_id__name"."name" = $2 AND "product_product__product_tmpl_id__name"."lang" = $3 AND "product_product__product_tmpl_id__name"."value" != $4),"product_pro |  303418.37 |   614 | 494.17 |          61.54
 SELECT "product_product".id FROM "product_product","product_template" as "product_product__product_tmpl_id" LEFT JOIN "ir_translation" as "product_product__product_tmpl_id__name" ON ("product_product__product_tmpl_id"."id" = "product_product__product_tmpl_id__name"."res_id" AND "product_product__product_tmpl_id__name"."type" = $1 AND "product_product__product_tmpl_id__name"."name" = $2 AND "product_product__product_tmpl_id__name"."lang" = $3 AND "product_product__product_tmpl_id__name"."value" ! |   99612.90 |   286 | 348.30 |          20.20
 SELECT "product_product".id FROM "product_product" LEFT JOIN "product_template" as "product_product__product_tmpl_id" ON ("product_product"."product_tmpl_id" = "product_product__product_tmpl_id"."id") LEFT JOIN "ir_translation" as "product_product__product_tmpl_id__name" ON ("product_product__product_tmpl_id"."id" = "product_product__product_tmpl_id__name"."res_id" AND "product_product__product_tmpl_id__name"."type" = $1 AND "product_product__product_tmpl_id__name"."name" = $2 AND "product_produ |   43074.52 |   614 |  70.15 |           8.74

```

**Comment by Yenthe666** [comment]
The issue has been found and was related to a scheduled action that slowed down everything.
